src/main/java/io/getunleash/repository/FeatureRepositoryImpl.java

```diff
@@ -0,0 +1,219 @@
+    private final Throttler throttler;
+    private boolean ready;
+
+    public FeatureRepositoryImpl(UnleashConfig unleashConfig, UnleashEngine engine) {
+        this(unleashConfig, new FeatureBackupHandlerFile(unleashConfig), engine);
+    }
+
+    public FeatureRepositoryImpl(
+            UnleashConfig unleashConfig, BackupHandler featureBackupHandler, UnleashEngine engine) {
+        this(
+                unleashConfig,
+                featureBackupHandler,
+                engine,
+                unleashConfig.getUnleashFeatureFetcherFactory().apply(unleashConfig));
+    }
+
+    public FeatureRepositoryImpl(
+            UnleashConfig unleashConfig,
+            BackupHandler featureBackupHandler,
+            UnleashEngine engine,
+            FeatureFetcher fetcher) {
+        this(
+                unleashConfig,
+                featureBackupHandler,
+                engine,
+                fetcher,
+                unleashConfig.getToggleBootstrapProvider());
+    }
+
+    public FeatureRepositoryImpl(
+            UnleashConfig unleashConfig,
+            BackupHandler featureBackupHandler,
+            UnleashEngine engine,
+            FeatureFetcher fetcher,
+            ToggleBootstrapProvider bootstrapHandler) {
+        this(
+                unleashConfig,
+                featureBackupHandler,
+                engine,
+                fetcher,
+                bootstrapHandler,
+                new EventDispatcher(unleashConfig));
+    }
+
+    public FeatureRepositoryImpl(
+            UnleashConfig unleashConfig,
+            BackupHandler featureBackupHandler,
+            UnleashEngine engine,
+            FeatureFetcher fetcher,
+            ToggleBootstrapProvider bootstrapHandler,
+            EventDispatcher eventDispatcher) {
+        this.unleashConfig = unleashConfig;
+        this.featureBackupHandler = featureBackupHandler;
+        this.engine = engine;
+        this.featureFetcher = fetcher;
+        this.bootstrapper = bootstrapHandler;
+        this.eventDispatcher = eventDispatcher;
+        this.throttler =
+                new Throttler(
+                        (int) unleashConfig.getFetchTogglesInterval(),
+                        300,
+                        unleashConfig.getUnleashURLs().getFetchTogglesURL());
+        this.initCollections(unleashConfig.getScheduledExecutor());
+    }
+
+    @SuppressWarnings("FutureReturnValueIgnored")
+    private void initCollections(UnleashScheduledExecutor executor) {
+        Optional<String> features = this.featureBackupHandler.read();
+        if (!features.isPresent() && this.bootstrapper != null) {
+            features = this.bootstrapper.read();
+        }
+        if (features.isPresent()) {
+            try {
+                this.engine.takeState(features.get());
+            } catch (YggdrasilInvalidInputException | YggdrasilError e) {
+                LOGGER.error("Error when initializing feature toggles", e);
```

Comment:
Should this emit an event?

Reply:
Added in https://github.com/Unleash/unleash-client-java/pull/265/commits/1331f3fcd009772a136aac0bbfa8321206b18b9a